src/controllers/orbit-controller/orbit-state.js

```diff
@@ -0,0 +1,377 @@
+        rotationX: newRotationX,
+        rotationY: newRotationY
+      }));
+      const newCenterPos = newViewport.project(startRotateCenter);
+
+      newTranslationX += oldCenterPos[0] - newCenterPos[0];
+      newTranslationY -= oldCenterPos[1] - newCenterPos[1];
+    }
+
+    return this._getUpdatedOrbitState({
+      rotationX: newRotationX,
+      rotationY: newRotationY,
+      translationX: newTranslationX,
+      translationY: newTranslationY
+    });
+  }
+
+  /**
+   * End rotating
+   * Must call if `rotateStart()` was called
+   */
+  rotateEnd() {
+    return this._getUpdatedOrbitState({
+      startRotateCenter: null,
+      startRotateViewport: null
+    });
+  }
+
+  /**
+   * Start zooming
+   * @param {[Number, Number]} pos - position on screen where the pointer grabs
+   */
+  zoomStart({pos}) {
+    return this._getUpdatedOrbitState({
+      startZoomPos: pos,
+      startZoom: this._viewportProps.zoom
+    });
+  }
+
+  /**
+   * Zoom
+   * @param {[Number, Number]} pos - position on screen where the current center is
+   * @param {[Number, Number]} startPos - the center position at
+   *   the start of the operation. Must be supplied of `zoomStart()` was not called
+   * @param {Number} scale - a number between [0, 1] specifying the accumulated
+   *   relative scale.
+   */
+  zoom({pos, startPos, scale}) {
+    const {zoom, minZoom, maxZoom, width, height, translationX, translationY} = this._viewportProps;
+
+    const startZoomPos = this._interactiveState.startZoomPos || startPos || pos;
+
+    const newZoom = clamp(zoom * scale, minZoom, maxZoom);
+    const deltaX = pos[0] - startZoomPos[0];
+    const deltaY = pos[1] - startZoomPos[1];
+
+    // Zoom around the center position
+    const cx = startZoomPos[0] - width / 2;
+    const cy = height / 2 - startZoomPos[1];
+    const newTranslationX = cx - (cx - translationX) * newZoom / zoom + deltaX;
+    const newTranslationY = cy - (cy - translationY) * newZoom / zoom - deltaY;
+
+    return this._getUpdatedOrbitState({
+      zoom: newZoom,
+      translationX: newTranslationX,
+      translationY: newTranslationY
+    });
+  }
+
+  /**
+   * End zooming
+   * Must call if `zoomStart()` was called
+   */
+  zoomEnd() {
+    return this._getUpdatedOrbitState({
+      startZoomPos: null,
+      startZoom: null
+    });
+  }
+
+  /* Private methods */
+
+  _getUpdatedOrbitState(newProps) {
+    // Update _viewportProps
+    return new OrbitState(Object.assign({}, this._viewportProps, this._interactiveState, newProps));
+  }
+
+  // Apply any constraints (mathematical or defined by _viewportProps) to map state
+  _applyConstraints(props) {
+    // Ensure zoom is within specified range
+    const {maxZoom, minZoom, zoom} = props;
+    props.zoom = zoom > maxZoom ? maxZoom : zoom;
+    props.zoom = zoom < minZoom ? minZoom : zoom;
+
+    return props;
+  }
+
+  /* Cast a ray into the screen center and take the average of all
```

Comment:
A lot going on here. Should this go into its own file, together with all the helper functions?

A ray casting algorithm would be a nice addition to the library. Maybe this could become a starting point for that?